Sources/App/Migrations/044/UpdateVersionReferenceNotNullable.swift

```diff
@@ -0,0 +1,44 @@
+// Copyright 2020-2021 Dave Verwer, Sven A. Schmidt, and other contributors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Fluent
+import SQLKit
+
+struct UpdateVersionReferenceNotNullable: AsyncMigration {
+    func prepare(on database: Database) async throws {
+        try await database.transaction { tx in
+            guard let db = tx as? SQLDatabase else {
+                fatalError("Database must be an SQLDatabase ('as? SQLDatabase' must succeed)")
+            }
+            try await db.raw(
+                #"DELETE FROM "versions" WHERE "reference" IS NULL"#
+            ).run()
```

Comment:
We don't currently have any rows `WHERE "reference" IS NULL` but there's technically nothing to prevent that. We therefore have a decision to make with respect to the reliability of the migration:

- alter the table and fail if there are rows that can't meet the new requirement
- ensure the migration will always succeed by fulfilling the precondition of no reference being NULL

In other cases where there's a clear default value (e.g. "score"), we run an update statement before the `ALTER TABLE` statement to set the field to 0 in case there are `NULL` values and we set a default on the column.

Here, there's not clear default. We could set `.branch("main")` but it's not a great default. I think it's better to reject saves without a reference entirely.

That also means we have no value to fill in for `NULL` fields (which we don't actually have). So instead I put in a `DELETE` which will effectively "run empty". The point is really to make the migration locally consistent and guarantee success even if somehow a version without a reference ended up in the table.

There is no harm deleting a version, especially one without a reference, which would be ignored by any of our processing, because if it was actually valid it would be recreated properly by the analysis stage.

Reply:
Sounds good.